File: parsers/gear.py
import json
import os

def del_if_exists(obj, prop):
    if prop in obj:
        del obj[prop]

def extract_from_system(obj, prop):
    if prop in obj["system"] and  isinstance(obj["system"][prop], int):
        obj[prop] = obj["system"][prop]

    elif prop in obj["system"] and obj["system"][prop] is not None:
        if "value" in obj["system"][prop] and prop != "traits" and prop != "damage" and prop != "hp":
            obj[prop] = obj["system"][prop]["value"]
        else:
            obj[prop] = obj["system"][prop]
        
        del obj["system"][prop]

input_dir = "input"
output_file = "output/output.json"
equipment = json.loads("[]")

print("processing...")
for filename in os.listdir(input_dir):
    f = os.path.join(input_dir,filename)
    fd = open(f, 'r')
    gear = json.load(fd)
    
    del_if_exists(gear,"_id")
    del_if_exists(gear,"img")
    del_if_exists(gear,"type")

    extract_from_system(gear, "armor")
    extract_from_system(gear, "baseItem")
    extract_from_system(gear, "category")
    extract_from_system(gear, "check")
    extract_from_system(gear, "damage")
    extract_from_system(gear, "dex")
    extract_from_system(gear, "description")
    extract_from_system(gear, "group")
    extract_from_system(gear, "hardness")
    extract_from_system(gear, "hp")
    extract_from_system(gear, "level")
    extract_from_system(gear, "price")
    extract_from_system(gear, "source")
    extract_from_system(gear, "strength")
    extract_from_system(gear, "speed")
    extract_from_system(gear, "traits")
    extract_from_system(gear, "usage")
    extract_from_system(gear, "weight")

    del gear["system"]
    equipment.append(gear)

equipment.sort(key = lambda s: s["name"])
output = open(output_file, 'w')
json.dump(equipment, output)
print("done")

File: ts/pf2e/gear.ts
import * as ko from "knockout";
import { CatalogObject, CatalogTraits } from "../shared/catalog";

export class Gear extends CatalogObject {
    acBonus: number;
    baseItem: string;
    category: string;
    checkPenalty: number;
    name: string;
    damage: DamageStats;
    description: string;
    dexCap: number;
    group: string;
    hardness: number;
    health: GearHealth;
    level: number;
    price: any;
    source: string;
    traits: CatalogTraits;
    usage: string;

    rarityFormatted: ko.PureComputed<string> = ko.pureComputed(() => {
        return this.capitalize(this.traits.rarity);
    });
    rarityClass: ko.PureComputed<string> = ko.pureComputed(() => {
        return "catalog-tag rarity-" + this.traits.rarity; 
    });
    tags: ko.PureComputed<string[]> = ko.pureComputed(() => {
	    return this.traits.value
            .map(x => this.capitalize(x))
            .sort();
    });    

    typeAndLevel: ko.PureComputed<string> = ko.pureComputed(() => {
        return "Item  " + this.level;
    });    
    
    sourceFormatted: ko.PureComputed<string> = ko.pureComputed(() => {
        return "Source: " + this.source;
    });

    constructor(json?: any){
        super();

        if(json === undefined){
            this.acBonus = 0;
            this.baseItem = "";
            this.category = "";
            this.checkPenalty = 0;
            this.name = "";

            this.damage = new DamageStats;
            this.damage.damageType = "";
            this.damage.dice = 0;
            this.damage.die = "";
            this.damage.value = "";

            this.dexCap = 0;
            this.description = "";
            this.group = "";
            this.hardness = 0;

            this.health = new GearHealth;
            this.health.brokenThreshold = 0;
            this.health.maxHP = 0;
            this.health.currentHP = 0;

            this.level = 1;
            this.price = { "": 0 }
            this.source = "";

            this.traits = new CatalogTraits;
            this.traits.custom = "";
            this.traits.rarity = "";
            this.traits.value = [];
            
            this.usage = "";
        }
        else{
            this.acBonus = json["armor"];
            this.baseItem = json["baseItem"];
            this.category = json["category"];
            this.checkPenalty = json["check"];
            this.name = json["name"];

            this.damage = new DamageStats;
            if (json["damage"]){
                this.damage.damageType = json["damage"]["damageType"];
                this.damage.dice = json["damage"]["dice"];
                this.damage.die = json["damage"]["die"];
                this.damage.value = json["damage"]["value"];
            }

            this.description = json["description"];
            this.dexCap = json["dex"];
            this.group = json["group"];
            this.hardness = json["hardness"];

            this.health = new GearHealth;
            if (json["hp"]){
                this.health.brokenThreshold = json["hp"]["brokenThreshold"];
                this.health.maxHP = json["hp"]["max"];
                this.health.currentHP = json["hp"]["value"];
            }

            this.level = json["level"] ? +json["level"] : 0;
            this.price = json["price"];
            this.source = json["source"];
            
            this.traits = new CatalogTraits;
            this.traits.custom = json["traits"]["custom"];
            this.traits.rarity = json["traits"]["rarity"];
            this.traits.value = json["traits"]["value"];

            this.usage = json["usage"];
        }
    }
}

class DamageStats {
    damageType: string;
    dice: number;
    die: string;
    value: string;  //unused?
}

class GearHealth {
    brokenThreshold: number;
    maxHP: number;
    currentHP: number;
}